### Zadanie z zajęć
1. Wyświetl liczebność w każdej z przedstawionych klas w stosunku do pierwszych 4 cech.
2. Wyświetl liczebność każdej z klas (class_type).
3. Wybierz trzy klasy : 1,2,7 oraz cechy takie jak 'hair', 'milk', 'aquatic', 'eggs': .
4. Podziel zbiór i wydziel: X_train, X_test, y_train, y_test
   parametry podziału: test_size=0.2, random_state=4
KNN:
1. Dopasuj dane (fit, predict)
2. Wyświetl dokładność klasyfikacji (accuracy_score), macierz pomyłek: confusion_matrix, raport z klasyfikacji: classification_report

In [18]:
import numpy as np
import pandas as pd
from collections import Counter
from IPython.display import display
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split


def train(X_train, y_train):
    return

def pred(X_train, y_train, x_test, k):
    distances = []
    targets = []

    for i in range(len(X_train)):
        distance =  np.linalg.norm(x_test - X_train[i,:])
        distances.append([distance, i])
        distances = sorted(distances)

    for i in range(k):
        index = distances[i][1]
        targets.append(y_train[index])
        return Counter(targets).most_common(1)[0][0]

def KNN(X_train, y_train, X_test, predictions, k):
    train(X_train, y_train)
    for i in range(len(X_test)):
        predictions.append(pred(X_train, y_train, X_test[i, :], k))
        
###############################################################################################################################

data=pd.read_csv('zoo.csv')
display(data.head())
df=pd.DataFrame(data)

data_1=df[['hair','feathers','eggs','milk','class_type']].groupby(['hair','feathers','eggs','milk']).count()
display(data_1)
counter=df.groupby('class_type').size()
print(counter,'\n')

data_2=df[(df['class_type']==1) | (df['class_type']==2) | (df['class_type']==7)]
data_2=data_2[['hair', 'milk', 'aquatic', 'eggs','class_type']]
print(data_2)

X = np.array(data_2.iloc[:, 0:4])
y = np.array(data_2['class_type'])
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.2,random_state=4)

predictions = []
KNN(X_train, y_train, X_test, predictions, 8)
print("\nAccuracy: ",accuracy_score(y_test, predictions))

neigh = KNeighborsClassifier(n_neighbors=8)
neigh.fit(X_train, y_train)
y_pred=neigh.predict(X_test)

print("\nAccuracy: ",accuracy_score(y_test,y_pred))
print("\nConfusion matrix: \n",confusion_matrix(y_test,y_pred))
print("\nClassification report: \n",classification_report(y_test,y_pred))

animal_name  hair  feathers  eggs  milk  airborne  aquatic  predator  \
0    aardvark     1         0     0     1         0        0         1   
1    antelope     1         0     0     1         0        0         0   
2        bass     0         0     1     0         0        1         1   
3        bear     1         0     0     1         0        0         1   
4        boar     1         0     0     1         0        0         1   

   toothed  backbone  breathes  venomous  fins  legs  tail  domestic  catsize  \
0        1         1         1         0     0     4     0         0        1   
1        1         1         1         0     0     4     1         0        1   
2        1         1         0         0     1     0     1         0        0   
3        1         1         1         0     0     4     0         0        1   
4        1         1         1         0     0     4     1         0        1   

   class_type  
0           1  
1           1  
2           4  
3           1  
4           1

class_type
hair feathers eggs milk            
0    0        0    0              2
                   1              2
              1    0             34
     1        1    0             20
1    0        0    1             38
              1    0              4
                   1              1

class_type
1    41
2    20
3     5
4    13
5     4
6     8
7    10
dtype: int64 

     hair  milk  aquatic  eggs  class_type
0       1     1        0     0           1
1       1     1        0     0           1
3       1     1        0     0           1
4       1     1        0     0           1
5       1     1        0     0           1
..    ...   ...      ...   ...         ...
95      0     0        0     1           2
96      1     1        0     0           1
98      1     1        0     0           1
99      0     0        0     1           7
100     0     0        0     1           2

[71 rows x 5 columns]

Accuracy:  0.9333333333333333

Accuracy:  0.9333333333333333

Confusion matrix: 
 [[11  0  0]
 [ 0  2  0]
 [ 0  1  1]]

Classification report: 
               precision    recall  f1-score   support

           1       1.00      1.00      1.00        11
           2       0.67      1.00      0.80         2
           7       1.00      0.50      0.67         2

    accuracy   

### Zadanie z zajęć c.d.
7.Zaimplementuj kolejno klasyfikatory:
- drzewa decyzyjne: tree.DecisionTreeClassifier(). W tym przypadku należy również wyznazyć istotność cech.
- random forest: RandomForestClassifier(n_estimators=100)
- regresję logistyczną: LogisticRegression(solver='lbfgs')

Do każdego klasyfikatora przedstaw metryki:
- dokładność: accuracy_score
- macierz pomyłek: confusion_matrix
- raport z klasyfikacji: classification_report

In [19]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

data_3=df[(df['class_type']==1) | (df['class_type']==2) | (df['class_type']==7)]
data_3=data_3[['hair', 'milk', 'aquatic', 'tail','class_type']]
X1 = np.array(data_3.iloc[:, 0:4])
y1 = np.array(data_3['class_type'])
X1_train, X1_test, y1_train, y1_test=train_test_split(X1,y1,test_size=0.2,random_state=4)

tree=DecisionTreeClassifier()
tree.fit(X1_train,y1_train)
importance_t=tree.feature_importances_
y1_pred=tree.predict(X1_test)

importance_tree=pd.DataFrame(data={'Features':['hair', 'milk', 'aquatic', 'tail'],
                                 'Importance': [value for i,value in enumerate(importance_t)]})

print("\n\033[1mDrzewo decyzyjne:\n\033[0m")
print(importance_tree)
print("\nAccuracy: ",accuracy_score(y1_test,y1_pred))
print("\nConfusion matrix: \n",confusion_matrix(y1_test,y1_pred))
print("\nClassification report: \n",classification_report(y1_test,y1_pred))

forest = RandomForestClassifier(n_estimators=100)
forest.fit(X1_train,y1_train)
y11_pred=forest.predict(X1_test)

print("\n\033[1mRandom Forest:\033[0m")
print("\nAccuracy: ",accuracy_score(y1_test,y11_pred))
print("\nConfusion matrix: \n",confusion_matrix(y1_test,y11_pred))
print("\nClassification report: \n",classification_report(y1_test,y11_pred))

log_regr=LogisticRegression(solver='lbfgs')
log_regr.fit(X1_train,y1_train)
y12_pred=log_regr.predict(X1_test)

print("\n\033[1mRegresja logistyczna:\033[0m")
print("\nAccuracy: ",accuracy_score(y1_test,y12_pred))
print("\nConfusion matrix: \n",confusion_matrix(y1_test,y12_pred))
print("\nClassification report: \n",classification_report(y1_test,y12_pred))


Drzewo decyzyjne:

  Features  Importance
0     hair    0.000000
1     milk    0.664336
2  aquatic    0.000000
3     tail    0.335664

Accuracy:  0.9333333333333333

Confusion matrix: 
 [[11  0  0]
 [ 0  2  0]
 [ 0  1  1]]

Classification report: 
               precision    recall  f1-score   support

           1       1.00      1.00      1.00        11
           2       0.67      1.00      0.80         2
           7       1.00      0.50      0.67         2

    accuracy                           0.93        15
   macro avg       0.89      0.83      0.82        15
weighted avg       0.96      0.93      0.93        15


Random Forest:

Accuracy:  0.9333333333333333

Confusion matrix: 
 [[11  0  0]
 [ 0  2  0]
 [ 0  1  1]]

Classification report: 
               precision    recall  f1-score   support

           1       1.00      1.00      1.00        11
           2       0.67      1.00      0.80         2
           7       1.00      0.50      0.67         2

    accuracy        

### Zadanie
1. Poszukaj danych do klasyfikacji na stronie kaggle.
2. Wczytaj je, wybierz dwie najbardziej liczne klasy.
3. Podziel dane na uczące i testowe.
4. Przeprowadź klasyfikację z wykorzystaniem algorytmu K-najbliższych sąsiadów, drzew decyzyjnych, lasów losowych oraz regresji logistyczne dla zdefiniowanych dwóch klas. Wyniki przedstaw w postaci: dokładność, macierz pomyłek oraz raportu klasyfikacji.
5. Zaznacz metodę najbardziej skuteczną (może być to w formie opisu, graficznie - styl dowolny).
6. Pamiętaj, że tworząc skrypt w Jupyter Notebook objaśniaj wprowadzane dane oraz zapisy, aby osoba czytająca taki raport (Wasz rodzic, kolega nie-będący-potencjalnym inżynierem/programistą) była w stanie zrozumieć ideę klasyfikacji (zaprezentuj dane, przedstaw ich liczebność, cechy brane do klasyfikacji), przeprowadzony proces klasyfikacji z wykorzystaniem kilku klasyfikatorów wraz z bardzo krótką ideą działania klasyfikatora o największej wartości dokładności klasyfikacji).

In [20]:
import numpy as np
import pandas as pd
from collections import Counter
from IPython.display import display
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import make_classification
from sklearn.metrics import classification_report
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split


data=pd.read_csv('pokemon.csv')
df=pd.DataFrame(data)
display(data.head())
count=df[['Type 1']].groupby('Type 1').size().sort_values(ascending=False)
print(count)

print('''\nJak widać powyżej - największą liczebność mają klasy Water i Normal, takie więc zostaną wybrane do dalszej analizy.\n''')
classes=df[(df['Type 1']=='Water')|(df['Type 1']=='Normal')]
df_limited=classes[['HP','Attack','Defense','Speed','Type 1']]

display(df_limited)

X=np.array(df_limited.iloc[:,0:4])
y=np.array(df_limited['Type 1'])
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.2,random_state=4)

neigh = KNeighborsClassifier(n_neighbors=8)
neigh.fit(X_train, y_train)
y_pred=neigh.predict(X_test)

print("\n\033[1mMetoda K najbliższych sąsiadów:\033[0m")
print("\nAccuracy: ",accuracy_score(y_test,y_pred))
print("\nConfusion matrix: \n",confusion_matrix(y_test,y_pred))
print("\nClassification report: \n",classification_report(y_test,y_pred))

tree=DecisionTreeClassifier(random_state=4)
tree.fit(X_train,y_train)
y1_pred=tree.predict(X_test)

print("\n\033[1mMetoda drzewa decyzyjnego:\033[0m")
print("\nAccuracy: ",accuracy_score(y_test,y1_pred))
print("\nConfusion matrix: \n",confusion_matrix(y_test,y1_pred))
print("\nClassification report: \n",classification_report(y_test,y1_pred))

forest = RandomForestClassifier(n_estimators=100,random_state=4)
forest.fit(X_train,y_train)         
y11_pred=forest.predict(X_test)

print("\n\033[1mMetoda Random Forest:\033[0m")
print("\nAccuracy: ",accuracy_score(y_test,y11_pred))
print("\nConfusion matrix: \n",confusion_matrix(y_test,y11_pred))
print("\nClassification report: \n",classification_report(y_test,y11_pred))

log_regr=LogisticRegression(solver='lbfgs')
log_regr.fit(X_train,y_train)
y12_pred=log_regr.predict(X_test)

print("\n\033[1mMetoda regresji logistycznej:\033[0m")
print("\nAccuracy: ",accuracy_score(y_test,y12_pred))
print("\nConfusion matrix: \n",confusion_matrix(y_test,y12_pred))
print("\nClassification report: \n",classification_report(y_test,y12_pred))

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

Type 1
Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Rock         44
Electric     44
Ground       32
Dragon       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
dtype: int64

Jak widać powyżej - największą liczebność mają klasy Water i Normal, takie więc zostaną wybrane do dalszej analizy.



HP  Attack  Defense  Speed  Type 1
9    44      48       65     43   Water
10   59      63       80     58   Water
11   79      83      100     78   Water
12   79     103      120     78   Water
20   40      45       40     56  Normal
..   ..     ...      ...    ...     ...
728  85      56       77     78  Normal
729  45      50       43     62  Normal
744  75      80       60    102  Normal
762  50      53       62     44   Water
763  71      73       88     59   Water

[210 rows x 5 columns]


Metoda K najbliższych sąsiadów:

Accuracy:  0.5714285714285714

Confusion matrix: 
 [[13  7]
 [11 11]]

Classification report: 
               precision    recall  f1-score   support

      Normal       0.54      0.65      0.59        20
       Water       0.61      0.50      0.55        22

    accuracy                           0.57        42
   macro avg       0.58      0.57      0.57        42
weighted avg       0.58      0.57      0.57        42


Metoda drzewa decyzyjnego:

Accuracy:  0.5238095238095238

Confusion matrix: 
 [[13  7]
 [13  9]]

Classification report: 
               precision    recall  f1-score   support

      Normal       0.50      0.65      0.57        20
       Water       0.56      0.41      0.47        22

    accuracy                           0.52        42
   macro avg       0.53      0.53      0.52        42
weighted avg       0.53      0.52      0.52        42


Metoda Random Forest:

Accuracy:  0.6904761904761905

Confusion matrix: 
 [[13  7]
 [ 6 16

### Która z metod jest najbardziej skuteczna?
Najbardziej skuteczną metodą jest Random Forest (po polsku: metoda lasów losowych). Jak widać powyżej - posiada ona najwyższą wartość dokładności (accuracy) obliczaną według wzoru: Accuracy = (TP+TN)/(P+N) = (13+16)/((13+6)+(16+7)) = 0.6904761904761905, gdzie P=TP+FN, N=TN+FP. Ma również najwyższe parametry precision (poprawne rozpoznanie w zakresie danej klasy), recall (rozpoznanie ilości elementów danej klasy), f1-score (miara dokładności testu w obrębie klas). Tak samo dla wyliczonych średnich dla tych parametrów - są one najwyższe ze wszystkich metod. W związku z tym wnioskuje się, że metoda Random Forest faktycznie jest najskuteczniejsza.
### Krótka idea działania metody lasów losowych
Random forest jest zespołową metodą uczenia maszynowego, polegającą na konstruowaniu wielu drzew decyzyjnych w czasie uczenia (każde z drzew budowane na innym, losowo wybranym pozbiorze zbioru danych uczących). Decyzja podejmowana "przez głosowanie" - tzn. dla danego zestawu cech wygrywa klasa, do której prowadziło najwięcej drzew. Lasy losowe poprawiają tendencję pojedynczych drzew decyzyjnych do nadmiernego dopasowywania się do danych treningowych, przeuczenia (ang. overfitting), aczkolwiek nie eliminują jej całkowicie.